docs: make Pytorch primitive public [DET-3204]

shiyuann opened the pull request:
## Description

This PR aims at showing what the interface, documentation, and examples look like when Pytorch Flexible Primitives is public.

## Test Plan

Use the CI test suite and internally dogfood the feature.

## Commentary (optional)

N/A

<!-- User-facing API changes need the "User-facing API Change" label -->

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

## Description

Lead with the intended commit body in this description field. For breaking changes, please include "BREAKING CHANGE:" at the beginning of your commit body. At minimum, this section should include a bracketed reference to the Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.

## Test Plan

Describe the situations in which you've tested your change, and/or a screenshot as appropriate. Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.

## Commentary (optional)

Use this section of your description to add context to the PR. Could be for particularly tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc. You may intentionally leave this section blank and remove the title.
--->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

justin-determined-ai commented on docs/reference/api/pytorch.txt:
```diff
@@ -144,6 +146,32 @@ To perform gradient clipping Determined provides two pre-made callback classes:
     :members:
 
+
+Migration guide
```
blocking: Since documentation is long-lived, I'm not certain a migration guide makes sense in this API reference; this documentation will only work for users who are both familiar with the old interface, and want to move over to the new interface. Instead, I think this document should be updated to reflect the current state of the API. However, a migration guide is a great idea: perhaps for the release note, or even a separate article just for folks who do want to upgrade.

shiyuann replied:
I see. Will only put it in release notes.

shiyuann replied:
According to the dogfooding feedback, the two styles of the API are confusing. I change that by grouping the functions/configurations that users need to specify into two lists and move the migration plan to a note snippet. It might make sense to have a migration guide in the reference because we are supporting two styles of APIs in the same class. However, that note snipper might be a bit redundant.

rb-determined-ai commented on docs/tutorials/pytorch-mnist-tutorial.txt:
```diff
@@ -16,12 +16,10 @@ Overview
 To use a PyTorch model in Determined, you need to port the model to Determined's API. For most models, this porting process is straightforward, and once the model has been ported, all of the features of Determined will then be available: for example, you can do :ref:`distributed training <multi-gpu-training>` or :ref:`hyperparameter search <hyperparameter-tuning>` without changing your model code ...
 
-When training a PyTorch model, Determined provides a built-in training loop that feeds batches of data into your forward pass, performs backpropagation, and computes training metrics. Determined also handles checkpointing, log management, and device initialization. To plug your model code into the Determined training loop, you define methods to perform the following tasks:
+When training a PyTorch model, Determined provides a built-in training loop that feeds batches of data into your batch training function, where you can perform forward pass, backpropagation, and training metrics computation. Determined also handles checkpointing, log management, and device initialization. To plug your model code into the Determined training loop, you define methods to perform the ...
```
... that feeds each batch of data into your `train_batch` function, where you will perform the forward pass, the backpropagation, and the computation of training metrics for that batch.

rb-determined-ai commented on docs/tutorials/pytorch-mnist-tutorial.txt:
```diff
@@ -156,20 +137,25 @@ The last two methods we need to define are ``build_training_data_loader`` and ``
 
-Defining the Forward Pass
-"""""""""""""""""""""""""
+Defining the Training Function
```
Defining ``train_batch``

rb-determined-ai commented on docs/tutorials/pytorch-mnist-tutorial.txt:
```diff
@@ -156,20 +137,25 @@ The last two methods we need to define are ``build_training_data_loader`` and ``
+""""""""""""""""""""""""""""""
 
-The ``train_batch`` method computes the forward pass and calculates the loss for a single batch of data. Determined expects a dictionary with the calculated loss and other user-defined metrics and will automatically average all the metrics across batches. Since Determined handles the backpropagation call, we do not need to call ``loss.backwards()`` or ``optim.zero_grad()``.
+The ``train_batch`` method runs the forward passes on the losses and backward passes on the optimziers with a single batch of data. Determined expects a dictionary with the calculated loss and other user-defined metrics and will automatically average all the metrics across batches. Since ``step_optimizer`` handles the backpropagation call and zeroing out the gradients so there is no need to call  ...
```
> runs the forward passes on the losses and backward passes on the optimziers with a single batch of data

Reword this somehow.  In reality you run the backward pass on the loss to calculate the gradients, and you step the optimizer.  I'm not sure what the right phrasing is for how to describe the forward pass.  "forward pass through the model", maybe?

> Determined expects a dictionary with the calculated loss and other user-defined metrics

The loss is no longer special, right?

> ``step_optimizer`` handles the backpropagation

Is this true?

rb-determined-ai commented on docs/tutorials/pytorch-mnist-tutorial.txt:
```diff
@@ -156,20 +137,25 @@ The last two methods we need to define are ``build_training_data_loader`` and ``
         output = model(data)
         loss = torch.nn.functional.nll_loss(output, labels)
 
+        # Define the training backward pass and run optimization step.
+        self.context.backward(loss)
+        self.context.step_optimizer(self.optimizer)
```
non-blocking: split this comment into one comment per line of code here; I don't see a strong reason to fuse them in this example.

Also, instead of saying "run optimizer step", can we say "step optimizer"?

rb-determined-ai commented on docs/tutorials/pytorch-mnist-tutorial.txt:
```diff
@@ -37,24 +35,15 @@ Here is what the skeleton of our trial class looks like:
+            # Initialize the trial class and wrap the models, optimizers, and LR schedulers.
             pass
 
         def train_batch(self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int):
```
blocking: when we make this public, the examples should not include a `model` argument in `train_batch`

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -426,6 +442,24 @@ def _step_optimizer(
         is performed by the Determined training loop by setting the experiment configuration
         optimization.aggregation_frequency.
 
+        Here is a code example of using lambda function:
+
+        .. code-block:: python
+
+            self.context.step_optimizer(
+                self.opt1,
+                clip_grads=lambda params: torch.nn.utils.clip_grad_norm_(params, 0.0001),
+            )
+
+        Or you can use a function defined by ``def``:
+
+        .. code-block:: python
+
+            def clip_grads(params):
+                torch.nn.utils.clip_grad_norm_(params, 0.0001),
+
+            self.context.step_optimizer(self.opt1, clip_grads)
+
```
Question: do we need to explain how to pass functions to users?  Like could we just say "here's a code example" and give them the `def` version, and let users who know about `lambda` use lambda on their own?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -675,20 +691,75 @@ class PyTorchTrial(det.Trial):
+       We should pass a function into ``step_optimizer(optimizer, clip_grads=...)``
+       provided by `PyTorchTrialContext <pytorch-trial-context_>`_ in the :meth:`train_batch`.
+
+    Another deprecated method to do the above things is as follows:
+
+    1. Define the abstract methods :meth:`build_model`, :meth:`optimizer`,
+       :meth:`create_lr_scheduler` for the model, optimizer, and LR scheduler, which
+       initialized by Determined. Note this method doesn't support multiple models,
+       optimizers, and LR schedulers.
+
+    2. Configure the experiment configuration field ``optimization.mixed_precision`` for
+       automatic mixed precision.
+
+    2. Clip gradients by adding a callback with ``on_before_optimizer_step`` in
+       :meth:`build_callbacks`.
```
Can we avoid even mentioning the deprecated interface here?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -675,20 +691,75 @@ class PyTorchTrial(det.Trial):
+       :meth:`build_callbacks`.
+
+
+    .. _pytorch-trial-migration-guide:
+
+    .. note::
+
+        The new method of defining a PyTorch trial is more powerful in allowing
+        multiple models, optimizers, and LR schedulers and more flexible in running
+        forward and backward passes on them in arbitrary orders.
+        You can only choose one of the two methods.
+
+        To migrate from the deprecated method to the new one, please
+        change the following places in your code:
+
+        1. Wrap models, optimizers, and LR schedulers in the :meth:`__init__` with the methods
+           ``Model``, ``Optimizer``, and ``LRScheduler`` that are provided by
+           `PyTorchTrialContext <pytorch-trial-context_>`_. At the same time,
+           remove the implementation of :meth:`build_model`, :meth:`optimizer`,
+           :meth:`create_lr_scheduler`.
+
+        2. If using automatic mixed precision (AMP), configure Apex AMP in the ``__init__``
+           with the method ``configure_apex_amp`` that is provided by
+           `PyTorchTrialContext <pytorch-trial-context_>`_.
+           At the same time, remove the experiment configuration ``optimizations.mixed_precision``.
+
+        3. Run backward passes on losses and step optimizers in :meth:`train_batch` with the methods
+           ``backward`` and ``step_optimizer`` provided by
+           `PyTorchTrialContext <pytorch-trial-context_>`_. Clip graidents by passing a function
+           to the ``clip_grads`` argument of ``step_optimizer``
+           while removing the ``PyTorchCallback`` counterpart in :meth:`build_callbacks`.
```
non-blocking: can we make this a separate doc from PyTorchTrial?  I guess I'm not sure where we would put it, which is why I'm marking this as non-blocking.

shiyuann replied:
Made it a separate section.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -698,60 +769,62 @@ def __init__(self, trial_context: PyTorchTrialContext) -> None:
 
-        Override this function to initialize any shared state between the
-        function implementations.
+        1. Initialize models, optimizers, and LR schedulers and wrap them with ``context.Model``,
```
A couple of points.  First, let's use the restructured text syntax for numbered list:
```
#. some text
#. more text
```

Second, this numbered list is missing some context as to what it represents conceptually.  Maybe change the first line to something like:

```
"""
Initialize a trial using the provided trial context.  The general steps are:

#. Initialize model(s) and wrap them with ``context.Model``.
#. Initialize optimizer(s) and wrap them with ``context.Optimizer``.
#. Initialize learning rate schedulers and wrap them with ``context.LRScheduler``.
#. If desired, wrap models and optimizer with ``context.configure_apex_amp`` 
   to use ``apex.amp`` for automatic mixed precision.

Here is a code example:
...
```

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
 
     @abstractmethod
     def train_batch(
         self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
```
blocking: `train_batch` must not accept a `model` argument when we make this interface public, since that is a relic of the old API.

shiyuann replied:
Removing the `model` is going to break user code. I don't think we should remove this as long as we decide to maintain the backward compatibility.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
+        Users should implement this function by doing the following things:
+        1. Run forward passes on the models.
+
+        2. Calculate the gradients with the losses with ``backward``.
```
... with `context.backward`.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
+
+        2. Calculate the gradients with the losses with ``backward``.
+
+        3. Call an optimization step for the optimizers with ``step_optimizer``. You can clip
```
... with `context.step_optimizer`.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
+        Train on one batch.
+
+        Users should implement this function by doing the following things:
+        1. Run forward passes on the models.
```
use `#.` notation

rb-determined-ai replied:
ok

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
+            self.lrs1.step()
+            self.lrs2.step()
+
+            return {"loss1": loss1, "loss2": loss2}
```
Question: Do you know what do we render in the UI if we don't return a key named "loss"?  Do we plot all the training metrics on one graph?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
+
+            return {"loss1": loss1, "loss2": loss2}
+
+        In the deprecated method, there are no need to call ``context.backward`` and
+        `context.step_optimizer`.
```
Remove mention of deprecated things

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
+        Arguments:
+            batch (Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor):
+                batch of data for training.
+            model (torch.nn.Module): (deprecated) the only model used for training.
+                If there are multiple models, return the first one.
```
ditto

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -848,8 +946,14 @@ def evaluate_batch(self, batch: TorchData, model: nn.Module) -> Dict[str, Any]:
+        Arguments:
+            batch (Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor):
+                batch of data for evaluating.
+            model (torch.nn.Module): (deprecated) the only model used for evaluating.
+                If there are multiple models, return the first one.
```
blocking: this also must not take a `model` argument in the API.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -762,47 +835,72 @@ def create_lr_scheduler(
-        # TODO(DET-3267): deprecate the model argument when releasing pytorch flexible primitives.
+        .. code-block:: python
+
+            # Assume having initialized models, optimizers, and LR schedulers (when
+            # using manual mode): self.model1, self.model2, self.opt1, self.opt2, and self.lrs1.
```
Reword to:
```
Assume two models, two optimizers, and two learning rate schedulers were initialized in ``__init__``.
```

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -874,8 +978,13 @@ def evaluate_full_dataset(
+
+        Arguments:
+            data_loader (torch.utils.data.DataLoader): data loader for evaluating.
+            model (torch.nn.Module): (deprecated) the only model used for evaluating.
+                If there are multiple models, return the first one.
```
blocking: this also must not take a `model` argument in the API.

shiyuann commented:
@rb-determined-ai IMO, as long as we maintain full backward compatibility, we should keep the `model` argument in the trial functions. If we don't maintain backward compatibility, we should remove `build_model`, `optimizer`, `create_lr_scheduler`, and etc. Although we decide to maintain two interfaces in one class, I don't think being somewhere in-between is the right way forward.

rb-determined-ai commented:
> IMO, as long as we maintain full backward compatibility, we should keep the model argument in the trial functions

@shiyuann, if we take that strategy then we have two options:

- Leave the `model` parameters forever, even though in the future they would not make any sense.  @davidhershey was very clear during dogfooding that he does not like this strategy.
- Introduce a breaking change to the interface when we deprecate the old interface, since we would take away the `model` parameters then.  This is like the worst of both worlds because it means that the `train_batch` function signature would not really make sense until we deprecate the old functions, and additionally we would have an additional breaking change in the future.

What I am proposing is that we call `train_batch` with a `model` parameter when we know the old interface is in use, and we call `train_batch` without the `model` parameter if we know the new interface is in use.  I do think that's a little odd but I think it's a much better way forward than either of the above two options.

rb-determined-ai commented on examples/official/trial/cifar10_cnn_pytorch/model_def.py:
```diff
@@ -68,41 +66,42 @@ def build_model(self) -> nn.Module:
         # call before loading the backbone weights.
-        reset_parameters(model)
-        return model
+        reset_parameters(self.model)
```
can we call `self.model.reset_parameters()` instead?

shiyuann replied:
same.

rb-determined-ai commented on harness/determined/pytorch/_callback.py:
```diff
@@ -85,28 +91,38 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
 class ClipGradsL2Norm(PyTorchCallback):
     """
-    Callback that performs gradient clipping using
+    (Deprecated) Callback that performs gradient clipping using
```
Can we use `.. warning::` syntax for deprecation warnings rather than `(Deprecated)`?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -197,12 +209,18 @@ def _init_device(self) -> None:
         )
 
-    def _configure_apex_amp(
+    def configure_apex_amp(
```
non-blocking: do you think `apex_amp_initialize` is a better name than `configure_apex_amp`?  I kind of like it since the arguments are the same as `apex.amp.initialize()`.

shiyuann replied:
I guess `configure_apex_amp` sounds closer to natural language flow. so affirm the original judgment?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -412,26 +430,46 @@ def _average_gradients(self, parameters: Any, divisor: int) -> None:
             optimizer(``torch.optim.Optimizer``): Which optimizer should be stepped.
             clip_grads(a function, optional): This function should have one argument for
                 parameters in order to clip the gradients
+            auto_zero_grads(bool, optional): Automatically zero out gradients automatically
+                after stepping the optimizer. If set to be false, you need to call
+                ``optimizer.zero_grad()`` manually. Note that if
+                ``optimizations.aggregation_frequency`` is larger than 1, you can only
+                set it to be true.
```
We need to explicitly state that the default is `True`.  Also, rephrase `If set to be false` -> `If false` and `you can only set it to be true` -> `auto_zero_grads must be true`

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -674,85 +706,93 @@ def _save(self, path: pathlib.Path) -> workload.Response:
+       Note that we support arbitrary numbers of models, optimizers, and LR schedulers
+       and arbitrary orders of running forward and backward passes.
+
+    2. **Configure automatic mixed precision**.
```
Numbering is wrong.  I'm ok using actual numbers instead of `#.` to make the python doctrings easier to read, but this is an example of why `#.` is a generally a good idea when docstring readability is not a concern.

As a sidenote, I think these should actually be bullet points and not numbered entries anyway.

shiyuann replied:
Changed it to bulleted list

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -763,47 +803,68 @@ def create_lr_scheduler(
+            torch.Tensor or Dict[str, Any]:
+                training metrics to return.
+        """
+        # TODO(DET-3262): remove the model argument.
```
Remove TODO

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -849,8 +910,12 @@ def evaluate_batch(self, batch: TorchData, model: nn.Module) -> Dict[str, Any]:
+                batch of data for evaluating.
         """
-        # TODO(DET-3267): deprecate the model argument when releasing pytorch flexible primitives.
+        # TODO(DET-3262): remove the model argument.
```
remove TODO

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -875,8 +938,11 @@ def evaluate_full_dataset(
+            data_loader (torch.utils.data.DataLoader): data loader for evaluating.
         """
-        # TODO(DET-3267): deprecate the model argument when releasing pytorch flexible primitives.
+        # TODO(DET-3262): remove the model argument.
```
remove TODO

shiyuann commented on docs/reference/api/pytorch.txt:
```diff
@@ -133,17 +135,50 @@ Then, implement the ``build_callbacks`` function in ``PyTorchTrial``:
+To migrate from the deprecated method to the new one, please
+change the following places in your code:
+
+1. Wrap models, optimizers, and LR schedulers in the :meth:`__init__` method with the
```
done

shiyuann commented on docs/tutorials/pytorch-mnist-tutorial.txt:
```diff
@@ -118,19 +105,13 @@ The ``build_model`` method returns a ``torch.Module`` or ``torch.Sequential`` ob
-""""""""""""""""""""""
-The ``optimizer`` method returns a ``torch.optim`` object. The MNIST model code uses ``torch.optim.Adadelta`` and we can continue to use that in our implementation of ``optimizer``.
+        ))
+        reset_parameters(self.model)
```
It actually calls `.reset_parameters()` on each item of `self.model.named_modules()`. So you basically need to write a loop to do that. I would leave it to https://determinedai.atlassian.net/browse/DET-3578

rb-determined-ai replied:
Talked over zoom, we'll leave this for a separate PR, but it would be good to fix this before the major api update goes out (that is, in this next release).

shiyuann commented on examples/experimental/trial/mnist_pytorch_multi_output/model_def.py:
```diff
@@ -97,13 +97,23 @@ def compute_loss(predictions: TorchData, labels: TorchData) -> torch.Tensor:
+
+        # If loading backbone weights, do not call reset_parameters() or
+        # call before loading the backbone weights.
+        reset_parameters(self.model)
```
same.

shiyuann commented on examples/experimental/trial/imagenet_nas_arch_pytorch/model_def.py:
```diff
@@ -40,6 +38,60 @@ def __init__(self, context: det.TrialContext) -> None:
+
+        # If loading backbone weights, do not call reset_parameters() or
+        # call before loading the backbone weights.
+        reset_parameters(self.model)
```
same.

shiyuann commented on examples/official/native/native_mnist_pytorch/model_def.py:
```diff
@@ -10,32 +10,46 @@
+
+        # If loading backbone weights, do not call reset_parameters() or
+        # call before loading the backbone weights.
+        reset_parameters(self.model)
```
same.

shiyuann commented:
@rb-determined-ai 

> This looks great. I think there is only one big ask that I have, and that is for us to have stronger separation of the deprecated pytorch interface from the new one. This was the most important feeback that we got during the dogfooding I think.
> Why not just have a separate reference page entirely named PyTorchTrial (Deprecated interface) or something? Then we could have a separate automodule that uses the :members: option to list out only the deprecated methods. It would also be a great place to put the migration guide.

I guess users can go to an old version of the documentation for the deprecated case. We have already agreed on not holding the deprecated trial so `automodule` wouldn't work. It also makes sense to move the migration guide only in the release plan.

rb-determined-ai commented:
spoke in person, we're going to use `:exclude-members:` to exclude deprecated PyTorchTrial functions, then we are going to just link to the old version of the documentation.

shiyuann commented:
@rb-determined-ai I removed all the things related to the deprecated interface, including some functions in `PytorchTrial` and `PytorchTrialContext`. Updated the gradient clipping section. Linked to old documentation hosted in `docs.determined.ai`.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -674,85 +706,95 @@ def _save(self, path: pathlib.Path) -> workload.Response:
     ) -> Optional[LRScheduler]:
         """
-        Create a learning rate scheduler for the trial given an instance of the
+        (Deprecated) Create a learning rate scheduler for the trial given an instance of the
```
There's still a bunch of these laying around, I think we should use the same `warning::` syantax as before.

shiyuann commented:
@rb-determined-ai fixed this issue.
